lib/rules/array-callback-return.js

```diff
@@ -188,15 +223,28 @@ module.exports = {
          * @returns {void}
          */
         function checkLastSegment(node) {
-            if (funcInfo.shouldCheck &&
-                funcInfo.codePath.currentSegments.some(isReachable)
-            ) {
+
+            let messageId = null;
+
+            if (funcInfo.shouldCheck) {
+                if (funcInfo.arrayMethodName === "forEach") {
+                    if (options.checkForEach &&
+                        !funcInfo.codePath.currentSegments.some(isReachable)
```

Comment:
@gabrieldrs just to check, this part is still work in progress? Asking to avoid delay from my side in the case that you're waiting for review.

If arrow functions with implicitly returned expressions are causing trouble, that can be checked from the `expression` flag: [estree#ArrowFunctionExpression](https://github.com/estree/estree/blob/master/es2015.md#arrowfunctionexpression)

Reply:
It's still work in progress, I just submitted an update fixing some of the comments, haven't had much time to go into this more complex part, but will do! Thanks for the reference, I will further investigate it

Reply:
No problem, thanks for working on this!

Hm... I think the changes you already submitted are actually the more complex part. The logic for `forEach` doesn't have to check code path segments at all. Just `return` statements (which is already done) and if the function is an arrow function that implicitly returns value (in which case its `expression` flag is `true`).

Reply:
Ok, so I had this check because I was thinking about cases like these:

```js 
foo.forEach(function(x) {
    return;
})

foo.forEach(function(x) {
    if (x) {
        x++;
    }
    return;
})

foo.forEach(function(x) {
    x+=10;
    return;
    x++;
})
```

Should this rule consider these cases to be valid? If so, then I agree that we should remove this check.

On the arrow function matter, you consider implicit returns like `x => x+1` to be valid or invalid for this rule? I think they should be considered valid, because users often write them to abbreviate things like:
```js
var y = 0;
var x = [1, 2, 3];

x.forEach(val => y += val);

```

If we consider this invalid, users would have to add a pair of curly braces around the method body just to avoid returning it, but I'm not sure if we should enforce this...

Reply:
I think a bare return should be valid.

Separately, i don’t think a concise arrow function should ever be used when the return value doesn’t matter, so even if by option I’d like to see this rule block them inside forEach.

Reply:
> Ok, so I had this check because I was thinking about cases like these:
> 
> ```js
> foo.forEach(function(x) {
>     return;
> })
> 
> foo.forEach(function(x) {
>     if (x) {
>         x++;
>     }
>     return;
> })
> 
> foo.forEach(function(x) {
>     x+=10;
>     return;
>     x++;
> })
> ```
> 
> Should this rule consider these cases to be valid? If so, then I agree that we should remove this check.

I believe all these cases should be valid by this rule. The rule (the option) should warn only when `forEach` callback returns a value, so `return;` should be okay.

There are other rules to warn when `return;` doesn't seem to make sense, such as `no-useless-return` or `no-unreachable`.

> On the arrow function matter, you consider implicit returns like `x => x+1` to be valid or invalid for this rule? I think they should be considered valid, because users often write them to abbreviate things like:
> 
> ```js
> var y = 0;
> var x = [1, 2, 3];
> 
> x.forEach(val => y += val);
> ```
> 
> If we consider this invalid, users would have to add a pair of curly braces around the method body just to avoid returning it, but I'm not sure if we should enforce this...

I think this should be invalid, and that we can consider adding an option to allow this later in a separate PR, based on user feedback.

Reply:
Implemented the changes, had to add the `ArrowFunctionExpression` check in the `checkLastSegment` method, as this case has no return statement.